packages/engine/paima-funnel/src/funnels/midnight/funnel.ts

```diff
@@ -0,0 +1,451 @@
+    }
+
+    const result: PresyncChainData[] = [];
+
+    let lastLogTime = Date.now(),
+      lastLogBlock = start;
+    // TOOD: apply confirmation depth here
+    while (block && midnightTimestampToSeconds(block.timestamp) < startingBlock.timestamp) {
+      const now = Date.now();
+      if (now - lastLogTime > 5000) {
+        console.log(
+          `Midnight funnel ${this.chainInfo.config.networkId}: presync #${lastLogBlock}-${block.height}`
+        );
+
+        lastLogTime = now;
+        lastLogBlock = block.height + 1;
+      }
+
+      this.cache.nextBlockHeight = block.height + 1;
+
+      const extensionDatums = this.extensionsFromBlock(ENV.SM_START_BLOCKHEIGHT, block);
+      if (extensionDatums.length) {
+        result.push({
+          caip2: this.caip2,
+          networkType: ConfigNetworkType.MIDNIGHT,
+          extensionDatums,
+        });
+      }
+
+      block = await this.fetchBlock(this.cache.nextBlockHeight);
+    }
+
+    console.log(
+      `Midnight funnel ${this.chainInfo.config.networkId}: finished presync #${start}-${this.cache.nextBlockHeight - 1}`
+    );
+
+    const baseData = await baseDataPromise;
+    baseData[this.caip2] = result;
+    return baseData;
+  }
+
+  public override async readData(blockHeight: number): Promise<ChainData[]> {
+    const baseData = await this.baseFunnel.readData(blockHeight);
+
+    if (baseData.length === 0) {
+      return baseData;
+    }
+
+    const result: {
+      blockNumber: number;
+      extensionDatums?: ChainDataExtensionDatum[];
+      internalEvents?: InternalEvent[];
+      submittedData?: SubmittedData[];
+    }[] = [];
+    let block = await this.waitForBlock(this.cache.nextBlockHeight);
+
+    for (const baseBlock of baseData) {
+      // Process all Midnight blocks that precede the base block.
+      // Break when we've seen a block in the relative future.
+      // TODO: apply confirmation depth here.
+      while (midnightTimestampToSeconds(block.timestamp) < baseBlock.timestamp) {
+        // The meat: process this one.
+        console.log(
+          `Midnight funnel ${this.chainInfo.config.networkId}: base #${baseBlock.blockNumber} / Midnight #${block.height}`
+        );
+
+        const extensionDatums = this.extensionsFromBlock(baseBlock.blockNumber, block);
+        if (extensionDatums.length) {
+          result.push({
+            blockNumber: baseBlock.blockNumber,
+            extensionDatums,
+          });
+        }
+
+        // Fetch next block.
+        this.cache.nextBlockHeight = block.height + 1;
+        block = await this.waitForBlock(this.cache.nextBlockHeight);
+      }
+    }
+
+    // Insert the internal MidnightLastBlock event into each ChainData.
+    const composed = composeChainData(baseData, result);
+    for (const chainData of composed) {
+      (chainData.internalEvents ??= []).push({
+        type: InternalEventType.MidnightLastBlock,
+        caip2: this.caip2,
+        block: this.cache.nextBlockHeight - 1,
```

Comment:
does this work correctly if there is an exception in the middle of processing the returned (composed) data? Or if there is a crash. In the case where there are multiple blocks in baseData.

Reply:
Good catch. I followed after the Mina and Carp examples but missed that they were actually supplying different heights for each output `ChainData`. Fixed.